iu_xray_data.py

```diff
@@ -0,0 +1,218 @@
+        token_tensor: Dim [batch size, max_sentence_num + 1, max_word_num]
+        stop_prob: Dimension [batch size, max_sentence_num + 1]
+    """
+    img_tensor, label, caption_token, \
+        max_word_num, sentence_num = zip(*data)
+
+    img_tensor = torch.stack(img_tensor, 0)
+    label = torch.stack(label, 0)
+
+    max_sentence_num = max(sentence_num)
+    max_word_num = max(max_word_num)
+    batch_size = len(caption_token)
+
+    # During training, we will iterate through the second dimension of
+    # the token_tensor, the "+1" enables us to train the stop controller
+    # to predict "Stop" when at the end of paragraph generation
+    token_tensor = torch.zeros([batch_size, max_sentence_num + 1, max_word_num])
+    stop_prob = torch.zeros([batch_size, max_sentence_num + 1])
+
+    for i, token in enumerate(caption_token):
+        for j, sentence in enumerate(token):
+            token_tensor[i, j, :len(sentence)] = torch.Tensor(sentence)
+            stop_prob[i][j] = len(sentence) > 0
+
+    return img_tensor, label, token_tensor, stop_prob.to(torch.long)
+
+class IU_XRay_DataSource(DataSource):
+    r"""Dataset website here: https://openi.nlm.nih.gov/
+    NOTE: For the image without findings, we set the default
+    findings to be "the lungs are normal. "
+
+    Args:
+        hparams (dict or HParams, optional): IU_XRay_DataSource hyperparameters.
+            Missing hyperparameters will be set to default values.
+            See :meth:`default_hparams` for the hyperparameter
+            structure and default values.
+                * img_root (str): directory to the image root
+                * label_path (str): directory to the txt file that contains
+                    ground truth tags. Each line: [key, label]. Note that
+                    {key}.png is the name of the corresponding parent image.
+                * text_root (str): directory to the text root
+                    (e.g. findings, impression)
+                * vocab_path (str): directory to the txt file that contains all
+                    the words for the vocabulary
+                * transforms (str): data augmentation methods for input images
+    """
+
+    def __init__(self, hparams):
+        self._hparams = HParams(hparams, self.default_hparams())
+        self.img_root = self._hparams.img_root
+        self.transforms = self.build_transform(self._hparams.transforms)
+        self.text_root = self._hparams.text_root
+        self.vocab = Vocab(self._hparams.vocab_path)
+        self.pathologies = self._hparams.pathologies
+
+    def __len__(self):
+        return len(os.listdir(self.text_root))
+
+    def __iter__(self):
+        for file_name in os.listdir(self.text_root):
+            yield self.__getitem__(file_name)
+
+    def __getitem__(self, file_name):
+        json_name = osp.join(self.text_root, file_name)
+        with open(json_name, 'r') as f:
+            items = list(DataPack.deserialize(f.read()))
```

Comment:
So this is the not usual way to use datapack. This may create unexpected errors. You are assuming the index of how items are stored in the DataPack.

Please use the `get` and `get_single` function correctly. https://asyml-forte.readthedocs.io/en/latest/code/data.html#forte.data.data_pack.DataPack.get

Reply:
Fixed.